{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "imports_20150407.ipynb:26: MatplotlibDeprecationWarning: \n",
      "The savefig.jpeg_quality rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.\n",
      "  \"outputs_hidden\": false\n"
     ]
    }
   ],
   "source": [
    "%run imports.ipynb\n",
    "from os.path import join, isdir, isfile\n",
    "import sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting sh\n",
      "  Using cached sh-1.14.1-py2.py3-none-any.whl (40 kB)\n",
      "Installing collected packages: sh\n",
      "Successfully installed sh-1.14.1\n"
     ]
    }
   ],
   "source": [
    "!pip install sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "root_dir = \"/home/jovyan/notebooks/Far_West/admixture\"\n",
    "wd = join(root_dir, \"preprocessed_data\", \"PLINK_v1.07\")\n",
    "sh.mkdir(\"-p\", wd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimate_gdist(pos, table):\n",
    "    which = np.searchsorted(table.pposition.tolist(), pos + 1)    \n",
    "    gap = pos - table.iloc[which - 1].pposition\n",
    "    rate = table.iloc[which - 1].rrate\n",
    "    return np.round(table.iloc[which - 1].gposition + ((gap/1e6) * rate), 7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def loadmap(path):\n",
    "    mapdata = pd.read_csv(path, sep=\"\\t\", index_col=None)\n",
    "    return mapdata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapfn = join(\"admixture/map/Ag_{chrom}.map\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from anhima import loc, gt\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_fn = \"samples.meta.txt\"\n",
    "dat = pd.read_csv(meta_fn, sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src_code</th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>site</th>\n",
       "      <th>contributor</th>\n",
       "      <th>contact</th>\n",
       "      <th>year</th>\n",
       "      <th>m_s</th>\n",
       "      <th>sex</th>\n",
       "      <th>n_sequences</th>\n",
       "      <th>mean_coverage</th>\n",
       "      <th>ebi_sample_acc</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AOcol</th>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFcol</th>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFgam</th>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIcol</th>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMgam</th>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FRgam</th>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAgam</th>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GHcol</th>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GHgam</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GM</th>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNcol</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNgam</th>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GQgam</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GW</th>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>0</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KE</th>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UGgam</th>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            src_code  country  location  site  contributor  contact  year  \\\n",
       "population                                                                  \n",
       "AOcol             78       78        78    78           78       78    78   \n",
       "BFcol             75       75        75    75           75       75    75   \n",
       "BFgam             92       92        92    92           92       92    92   \n",
       "CIcol             71       71        71    71           71       71    71   \n",
       "CMgam            297      297       297   297          297      297   297   \n",
       "FRgam             24       24        24    24           24       24    24   \n",
       "GAgam             69       69        69    69           69       69    69   \n",
       "GHcol             55       55        55    55           55       55    55   \n",
       "GHgam             12       12        12    12           12       12    12   \n",
       "GM                65       65        65    65           65       65    65   \n",
       "GNcol              4        4         4     4            4        4     4   \n",
       "GNgam             40       40        40    40           40       40    40   \n",
       "GQgam              9        9         9     9            9        9     9   \n",
       "GW                91       91        91    91           91       91    91   \n",
       "KE                48       48        48    48           48       48    48   \n",
       "UGgam            112      112       112   112          112      112   112   \n",
       "\n",
       "            m_s  sex  n_sequences  mean_coverage  ebi_sample_acc  latitude  \\\n",
       "population                                                                   \n",
       "AOcol        78   78           78             78              78        78   \n",
       "BFcol        75   75           75             75              75        75   \n",
       "BFgam        92   92           92             92              92        92   \n",
       "CIcol        71   71           71             71              71        71   \n",
       "CMgam       297  297          297            297             297       297   \n",
       "FRgam        24   24           24             24              24        24   \n",
       "GAgam        69   69           69             69              69        69   \n",
       "GHcol        55   55           55             55              55        55   \n",
       "GHgam        12   12           12             12              12        12   \n",
       "GM            0   65           65             65              65        65   \n",
       "GNcol         4    4            4              4               4         4   \n",
       "GNgam        40   40           40             40              40        40   \n",
       "GQgam         9    9            9              9               9         9   \n",
       "GW            0   91           91             91              91        91   \n",
       "KE            0   48           48             48              48        48   \n",
       "UGgam       112  112          112            112             112       112   \n",
       "\n",
       "            longitude  \n",
       "population             \n",
       "AOcol              78  \n",
       "BFcol              75  \n",
       "BFgam              92  \n",
       "CIcol              71  \n",
       "CMgam             297  \n",
       "FRgam              24  \n",
       "GAgam              69  \n",
       "GHcol              55  \n",
       "GHgam              12  \n",
       "GM                 65  \n",
       "GNcol               4  \n",
       "GNgam              40  \n",
       "GQgam               9  \n",
       "GW                 91  \n",
       "KE                 48  \n",
       "UGgam             112  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat.groupby(\"population\").count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src_code</th>\n",
       "      <th>population</th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>site</th>\n",
       "      <th>contributor</th>\n",
       "      <th>contact</th>\n",
       "      <th>year</th>\n",
       "      <th>m_s</th>\n",
       "      <th>sex</th>\n",
       "      <th>n_sequences</th>\n",
       "      <th>mean_coverage</th>\n",
       "      <th>ebi_sample_acc</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ox_code</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AA0040-C</th>\n",
       "      <td>Twifo_Praso__E2</td>\n",
       "      <td>GHcol</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Twifo Praso</td>\n",
       "      <td>Twifo Praso</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>95033368</td>\n",
       "      <td>30.99</td>\n",
       "      <td>ERS311878</td>\n",
       "      <td>5.60858</td>\n",
       "      <td>-1.54926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA0041-C</th>\n",
       "      <td>Twifo_Praso__H3</td>\n",
       "      <td>GHcol</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Twifo Praso</td>\n",
       "      <td>Twifo Praso</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>95843804</td>\n",
       "      <td>31.70</td>\n",
       "      <td>ERS311886</td>\n",
       "      <td>5.60858</td>\n",
       "      <td>-1.54926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA0042-C</th>\n",
       "      <td>Takoradi_C7</td>\n",
       "      <td>GHcol</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Takoradi</td>\n",
       "      <td>Takoradi</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>107420666</td>\n",
       "      <td>35.65</td>\n",
       "      <td>ERS311894</td>\n",
       "      <td>4.91217</td>\n",
       "      <td>-1.77397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA0043-C</th>\n",
       "      <td>Takoradi_H8</td>\n",
       "      <td>GHcol</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Takoradi</td>\n",
       "      <td>Takoradi</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>95993752</td>\n",
       "      <td>29.46</td>\n",
       "      <td>ERS311902</td>\n",
       "      <td>4.91217</td>\n",
       "      <td>-1.77397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA0044-C</th>\n",
       "      <td>Takoradi_D10</td>\n",
       "      <td>GHcol</td>\n",
       "      <td>Ghana</td>\n",
       "      <td>Takoradi</td>\n",
       "      <td>Takoradi</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>103044262</td>\n",
       "      <td>33.67</td>\n",
       "      <td>ERS311910</td>\n",
       "      <td>4.91217</td>\n",
       "      <td>-1.77397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AY0087-C</th>\n",
       "      <td>Tia_dPM_46</td>\n",
       "      <td>CIcol</td>\n",
       "      <td>Cote d'Ivoire</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>101458232</td>\n",
       "      <td>32.56</td>\n",
       "      <td>ERS311822</td>\n",
       "      <td>5.89839</td>\n",
       "      <td>-4.82293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AY0088-C</th>\n",
       "      <td>Tia_dPM_52</td>\n",
       "      <td>CIcol</td>\n",
       "      <td>Cote d'Ivoire</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>95140584</td>\n",
       "      <td>30.75</td>\n",
       "      <td>ERS311830</td>\n",
       "      <td>5.89839</td>\n",
       "      <td>-4.82293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AY0089-C</th>\n",
       "      <td>Tia_aPM_4</td>\n",
       "      <td>CIcol</td>\n",
       "      <td>Cote d'Ivoire</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>92383724</td>\n",
       "      <td>29.04</td>\n",
       "      <td>ERS311838</td>\n",
       "      <td>5.89839</td>\n",
       "      <td>-4.82293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AY0090-C</th>\n",
       "      <td>Tia_aPM_13</td>\n",
       "      <td>CIcol</td>\n",
       "      <td>Cote d'Ivoire</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>96210618</td>\n",
       "      <td>30.94</td>\n",
       "      <td>ERS311846</td>\n",
       "      <td>5.89839</td>\n",
       "      <td>-4.82293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AY0091-C</th>\n",
       "      <td>Tia_aPM_37</td>\n",
       "      <td>CIcol</td>\n",
       "      <td>Cote d'Ivoire</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>David Weetman</td>\n",
       "      <td>2012</td>\n",
       "      <td>M</td>\n",
       "      <td>F</td>\n",
       "      <td>95973872</td>\n",
       "      <td>31.30</td>\n",
       "      <td>ERS311854</td>\n",
       "      <td>5.89839</td>\n",
       "      <td>-4.82293</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>505 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 src_code population        country     location         site  \\\n",
       "ox_code                                                                         \n",
       "AA0040-C  Twifo_Praso__E2      GHcol          Ghana  Twifo Praso  Twifo Praso   \n",
       "AA0041-C  Twifo_Praso__H3      GHcol          Ghana  Twifo Praso  Twifo Praso   \n",
       "AA0042-C      Takoradi_C7      GHcol          Ghana     Takoradi     Takoradi   \n",
       "AA0043-C      Takoradi_H8      GHcol          Ghana     Takoradi     Takoradi   \n",
       "AA0044-C     Takoradi_D10      GHcol          Ghana     Takoradi     Takoradi   \n",
       "...                   ...        ...            ...          ...          ...   \n",
       "AY0087-C       Tia_dPM_46      CIcol  Cote d'Ivoire     Tiassale     Tiassale   \n",
       "AY0088-C       Tia_dPM_52      CIcol  Cote d'Ivoire     Tiassale     Tiassale   \n",
       "AY0089-C        Tia_aPM_4      CIcol  Cote d'Ivoire     Tiassale     Tiassale   \n",
       "AY0090-C       Tia_aPM_13      CIcol  Cote d'Ivoire     Tiassale     Tiassale   \n",
       "AY0091-C       Tia_aPM_37      CIcol  Cote d'Ivoire     Tiassale     Tiassale   \n",
       "\n",
       "            contributor        contact  year m_s sex  n_sequences  \\\n",
       "ox_code                                                             \n",
       "AA0040-C  David Weetman  David Weetman  2012   M   F     95033368   \n",
       "AA0041-C  David Weetman  David Weetman  2012   M   F     95843804   \n",
       "AA0042-C  David Weetman  David Weetman  2012   M   F    107420666   \n",
       "AA0043-C  David Weetman  David Weetman  2012   M   F     95993752   \n",
       "AA0044-C  David Weetman  David Weetman  2012   M   F    103044262   \n",
       "...                 ...            ...   ...  ..  ..          ...   \n",
       "AY0087-C  David Weetman  David Weetman  2012   M   F    101458232   \n",
       "AY0088-C  David Weetman  David Weetman  2012   M   F     95140584   \n",
       "AY0089-C  David Weetman  David Weetman  2012   M   F     92383724   \n",
       "AY0090-C  David Weetman  David Weetman  2012   M   F     96210618   \n",
       "AY0091-C  David Weetman  David Weetman  2012   M   F     95973872   \n",
       "\n",
       "          mean_coverage ebi_sample_acc  latitude  longitude  \n",
       "ox_code                                                      \n",
       "AA0040-C          30.99      ERS311878   5.60858   -1.54926  \n",
       "AA0041-C          31.70      ERS311886   5.60858   -1.54926  \n",
       "AA0042-C          35.65      ERS311894   4.91217   -1.77397  \n",
       "AA0043-C          29.46      ERS311902   4.91217   -1.77397  \n",
       "AA0044-C          33.67      ERS311910   4.91217   -1.77397  \n",
       "...                 ...            ...       ...        ...  \n",
       "AY0087-C          32.56      ERS311822   5.89839   -4.82293  \n",
       "AY0088-C          30.75      ERS311830   5.89839   -4.82293  \n",
       "AY0089-C          29.04      ERS311838   5.89839   -4.82293  \n",
       "AY0090-C          30.94      ERS311846   5.89839   -4.82293  \n",
       "AY0091-C          31.30      ERS311854   5.89839   -4.82293  \n",
       "\n",
       "[505 rows x 15 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat = dat[dat.population != 'AOcol'] \n",
    "dat = dat[dat.population != 'CMgam'] \n",
    "dat = dat[dat.population != 'FRgam'] \n",
    "dat = dat[dat.population != 'GAgam'] \n",
    "dat = dat[dat.population != 'GQgam'] \n",
    "dat = dat[dat.population != 'KE'] \n",
    "dat = dat[dat.population != 'UGgam'] \n",
    "dat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "dat.to_csv('meta.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "sex_d = {\"F\": \"2\", \"M\": \"1\"}\n",
    "fstem = \"{chrom}_{desc}_{start}_{stop}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "callset_fn = callset_biallel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "script_d = join(wd, \"_script\")\n",
    "log_d = join(wd, \"_log\")\n",
    "sh.mkdir(\"-p\", script_d, log_d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_cmd = \"\"\"\n",
    "#! /bin/bash\n",
    "\n",
    "set -e \n",
    "set -o pipefail\n",
    "\n",
    "cd {wd}\n",
    "gunzip {file}.ped.gz {file}.map.gz\n",
    "plink --noweb --file {file} --out {file}.ld --indep-pairwise {nsnps} {slide} {r2}\n",
    "\n",
    "shuf -n 200000 {file}.ld.prune.in | sort > {file}.ld.prune.in.downsample\n",
    "\n",
    "plink --noweb --file {file} --extract {file}.ld.prune.in.downsample \\\\\n",
    "  --out {file}.ld.pruned --make-bed\n",
    "plink --noweb --file {file} --extract {file}.ld.prune.in.downsample \\\\\n",
    "  --out {file}.ld.pruned --recode\n",
    "\n",
    "gzip {file}.ped {file}.map\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "maf_min = 0.01\n",
    "r2_value = 0.1\n",
    "downsample_n = 100000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "regions = ((\"3R\", 1000000, 37000000, \"free\"),\n",
    "           (\"3L\", 15000000, 24_000_000, \"free\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True,  True,  True, ...,  True,  True,  True])"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_selection = metadata.population.isin({'GHcol', 'GHgam', 'BFgam', 'BFcol', 'GM', 'GW', 'GNgam', 'GNcol',\n",
    "       'CIcol'}).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "fh = callset_fn['3R']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import compress"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "fh_samples = [str(s) for s in callset_fn['3R'][\"samples\"][:]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "505"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(compress(fh_samples, pop_selection)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"allel allel-DisplayAs2D\"><span>&lt;GenotypeChunkedArray shape=(10752701, 505, 2) dtype=int8 chunks=(5251, 505, 2)\n",
       "   nbytes=10.1G cbytes=472.9M cratio=21.9\n",
       "   compression=blosc compression_opts={'cname': 'lz4', 'clevel': 5, 'shuffle': 1, 'blocksize': 0}\n",
       "   values=zarr.core.Array&gt;</span><table><thead><tr><th></th><th style=\"text-align: center\">0</th><th style=\"text-align: center\">1</th><th style=\"text-align: center\">2</th><th style=\"text-align: center\">3</th><th style=\"text-align: center\">4</th><th style=\"text-align: center\">...</th><th style=\"text-align: center\">500</th><th style=\"text-align: center\">501</th><th style=\"text-align: center\">502</th><th style=\"text-align: center\">503</th><th style=\"text-align: center\">504</th></tr></thead><tbody><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">0</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">1</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">2</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">...</th><td style=\"text-align: center\" colspan=\"12\">...</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">10752698</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">10752699</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr><tr><th style=\"text-align: center; background-color: white; border-right: 1px solid black; \">10752700</th><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">...</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td><td style=\"text-align: center\">0/0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "<GenotypeChunkedArray shape=(10752701, 505, 2) dtype=int8 chunks=(5251, 505, 2)\n",
       "   nbytes=10.1G cbytes=472.9M cratio=21.9\n",
       "   compression=blosc compression_opts={'cname': 'lz4', 'clevel': 5, 'shuffle': 1, 'blocksize': 0}\n",
       "   values=zarr.core.Array>"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g = allel.GenotypeChunkedArray(fh['calldata/GT']).subset(sel1=pop_selection)\n",
    "g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3R free\n",
      "Downsampling... to 100000 from 1313792\n",
      "3L free\n",
      "Downsampling... to 100000 from 327307\n"
     ]
    }
   ],
   "source": [
    "for chrom, start, stop, description in regions:\n",
    "    print(chrom, description)\n",
    "    fh = callset_fn[chrom]\n",
    "    fh_samples = [str(s) for s in callset_fn[chrom][\"samples\"][:]]\n",
    "    fh_samples = list(compress(fh_samples, pop_selection))\n",
    "\n",
    "    region_stem = fstem.format(chrom=chrom, start=start, stop=stop, desc=description)\n",
    "\n",
    "    # plink command\n",
    "    cmd = plink_cmd.format(file=region_stem, wd=wd, nsnps=500, slide=100, r2=r2_value)\n",
    "    sfn = join(script_d, \"PLINK_\" + region_stem + \".sh\")\n",
    "    with open(sfn, \"w\") as sf:\n",
    "        print(cmd, file=sf)\n",
    "    \n",
    "    positions = fh['variants']['POS'][:]\n",
    "    loci = (positions >= start) & (positions <= stop)\n",
    "    \n",
    "    g = allel.GenotypeChunkedArray(fh['calldata/GT']).compress(loci, axis=0).subset(sel1=pop_selection)\n",
    "        \n",
    "    pos = np.compress(loci, positions)\n",
    "    positions = None\n",
    "    \n",
    "    alleles = g.count_alleles()\n",
    "    \n",
    "    biallelic = alleles.max_allele() <= 1\n",
    "    g = g.compress(biallelic, axis=0)\n",
    "    pos = np.compress(np.array(biallelic), pos)\n",
    "    \n",
    "    freqs = g.count_alleles().to_frequencies()\n",
    "    maf_ok = np.min(np.array(freqs), axis=1) >= maf_min\n",
    "    g = g.compress(maf_ok, axis=0)\n",
    "    pos = np.compress(maf_ok, pos, axis=0)\n",
    "\n",
    "    if pos.size > downsample_n:\n",
    "        print(\"Downsampling... to {0} from {1}\".format(downsample_n, pos.size))\n",
    "        idx = np.random.choice(np.arange(0, pos.size), downsample_n, False)\n",
    "        idx.sort()\n",
    "        pos = np.take(pos, idx)\n",
    "        g = g.take(idx, axis=0)\n",
    "    \n",
    "    # create the ped file\n",
    "    fn = join(wd, region_stem + \".ped.gz\")\n",
    "    if not isfile(fn + \".ok\"):\n",
    "        with gzip.open(fn, \"wb\") as gz:\n",
    "            for i, sid in enumerate(fh_samples):\n",
    "                sex = sex_d[dat.loc[sid].sex]\n",
    "                if (chrom == \"X\") and (sex == \"1\"):\n",
    "                    continue\n",
    "                geno = np.array(g[:, i] + 1)\n",
    "                geno_str = \" \".join(np.apply_along_axis(\" \".join, 1, geno.astype(\"str\")))\n",
    "                line = \" \".join([sid, sid, \"0\", \"0\", sex, \"0\", geno_str]) + \"\\n\"\n",
    "                gz.write(line.encode())\n",
    "        \n",
    "        sh.touch(join(wd, fn + \".ok\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
